imap_processing/codice/codice_l1a_de.py

```diff
@@ -0,0 +1,468 @@
+    """
+    grouped_data = []  # Holds the properly grouped data to be decompressed
+    current_group = bytearray()  # Temporary storage for current group
+    group_byte_count = None  # Temporary storage for current group byte count
+
+    for packet_index in range(len(packets.event_data.data)):
+        packet_data = packets.event_data.data[packet_index]
+        group_code = packets.seq_flgs.data[packet_index]
+        byte_count = packets.byte_count.data[packet_index]
+
+        # If the group code is 3, this means the data is unsegmented
+        # and can be decompressed as-is
+        if group_code == SegmentedPacketOrder.UNSEGMENTED:
+            grouped_data.append(packet_data[:byte_count])
+
+        # If the group code is 1, this means the data is the first data in a
+        # group. Also, set the byte count for the group
+        elif group_code == SegmentedPacketOrder.FIRST_SEGMENT:
+            group_byte_count = byte_count
+            current_group += packet_data
+
+        # If the group code is 0, this means the data is part of the middle of
+        # the group.
+        elif group_code == SegmentedPacketOrder.CONTINUATION_SEGMENT:
+            current_group += get_de_metadata(packets, packet_index)
+            current_group += packet_data
+
+        # If the group code is 2, this means the data is the last data in the
+        # group
+        elif group_code == SegmentedPacketOrder.LAST_SEGMENT:
+            current_group += get_de_metadata(packets, packet_index)
+            current_group += packet_data
+
+            # The grouped data is now ready to be decompressed
+            values_to_decompress = current_group[:group_byte_count]
+            grouped_data.append(values_to_decompress)
+
+            # Reset the current group
+            current_group = bytearray()
+            group_byte_count = None
+
+    return grouped_data
+
+
+def unpack_bits(bit_structure: list, bit_vars: list, de_data: np.ndarray) -> dict:
+    """
+    Unpack 64-bit values into separate fields based on bit structure.
+
+    Parameters
+    ----------
+    bit_structure : list
+        List of bit lengths for each variable.
+    bit_vars : list
+        List of variable names to unpack.
+    de_data : np.ndarray
+        1D array of 64-bit values to unpack.
+
+    Returns
+    -------
+    dict
+        Dictionary of field_name -> unpacked values array.
+    """
+    unpacked = {}
+    # Data need to be unpacked in right to left order (LSB). Eg.
+    #   binary string  - 0x03 → 00000011
+    #   bit read order - Bit 7 → 0
+    #                    Bit 6 → 0
+    #                    Bit 5 → 0
+    #                    Bit 4 → 0
+    #                    Bit 3 → 0
+    #                    Bit 2 → 0
+    #                    Bit 1 → 1
+    #                    Bit 0 (LSB) → 1
+    #   bits chunks - [5, 1, ...., 7, 3, 16]
+    #   vars - ['gain', 'apd_id', ...., 'energy_step', 'priority', 'spare']
+    #   unpack data - [3, 0, 0, ....., 0, 0]
```

Comment:
Interesting, this is backwards of many other versions where the "spare" are the 0s on the right-hand-side so it is aligned with your array. Here your array is kind of backwards in a sense. I guess you could iterate through your inputs in reverse order and start at the far left edge if that woudl make anything easier too. But what you have seems OK, just slightly confusing at first glance because I'm not used to it.